[CORE-13254] ct: calculate global gc eligible L0 epoch

dotnwat opened the pull request:
Computes the top-level safe to delete L0 GC epoch value. There is a very simple ducktape test that verifies that deletes are occuring. I want to get some better testing into the ducktap setup for 25.3.x. I struggled a lot with figuring out how to get a fixture test working, so that's also on the todo list, but may not really be necessary for 25.3.x--Oren is going to take over getting this across the GA line.

Fixes: https://redpandadata.atlassian.net/browse/CORE-13254
Fixes: https://redpandadata.atlassian.net/browse/CORE-14824

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat committed:
ct: plumb topics table into level zero garbage collector

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ea3d38ca-afa6-465a-9751-81565f08c19e-->
<!--TEST RESULTS-->
<!--58188450-c89b-4f78-8a8b-a27cbcfee132-->
<!--TEST RESULTS-->
<!--faa238a4-2d98-4e2a-8877-b87022550e2d-->
<!--TEST RESULTS-->
<!--f957abab-8a90-4b4f-9c6c-369da5e1f4af-->
<!--TEST RESULTS-->
<!--3753e59e-d765-47d4-9671-b09e3f826289-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75616">build#75616</a></summary>


| test_class                                   | test_method                                  | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:---------------------------------------------|:---------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests                | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a51ce-6be0-47e6-aece-772084f8160d | FLAKY         | 20/21    | upstream reliability is '92.46298788694482'. current run reliability is '95.23809523809523'. drift is -2.77511 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming    |
| ConsumerGroupBalancingTest                   | test_coordinator_nodes_balance               | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a520a-46e7-43b8-9832-ca8dc38fda11 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupBalancingTest&test_method=test_coordinator_nodes_balance  |
| MountUnmountIcebergTest                      | test_simple_remount                          | {"cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a520a-46e7-43b8-9832-ca8dc38fda11 | FLAKY         | 17/21    | upstream reliability is '97.68421052631578'. current run reliability is '80.95238095238095'. drift is 16.73183 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
| SegmentMsTest                                | test_segment_rolling_with_retention_consumer | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a520a-46e0-4f42-95ac-5a05a7aad866 | FLAKY         | 16/21    | upstream reliability is '94.73039215686273'. current run reliability is '76.19047619047619'. drift is 18.53992 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| ShadowLinkingRandomOpsTest                   | test_node_operations                         | {"failures": false}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a520a-46e5-469d-b45a-9f86c076b39e | FLAKY         | 19/21    | upstream reliability is '99.69183359013869'. current run reliability is '90.47619047619048'. drift is 9.21564 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
| ShadowLinkingRandomOpsTest                   | test_node_operations                         | {"failures": true}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a520a-46e7-43b8-9832-ca8dc38fda11 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
| TxUpgradeCompactionTest                      | upgrade_with_compaction_test                 | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75616#019a520a-46e4-43cb-92f6-79e8606ebb10 | FLAKY         | 20/21    | upstream reliability is '99.3006993006993'. current run reliability is '95.23809523809523'. drift is 4.0626 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxUpgradeCompactionTest&test_method=upgrade_with_compaction_test       |
| src/v/storage/tests/storage_e2e_fixture_test | src/v/storage/tests/storage_e2e_fixture_test |                                                                              | unit        | https://buildkite.com/redpanda/redpanda/builds/75616#019a51b3-efb8-46e8-beee-bb3fc6525973 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77292">build#77292</a></summary>


| test_class            | test_method            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                               |
|:----------------------|:-----------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest | test_unavailable_nodes | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77292#019ae6ba-f6f3-41e8-aef3-6bcc290b5bbd | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_unavailable_nodes |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77379">build#77379</a></summary>


| test_class                    | test_method                           | test_arguments                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                | test_history                                                                                                                                                                 |
|:------------------------------|:--------------------------------------|:------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_topic_delete                     | {"source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/77379#019aecf0-a435-4228-bd79-047f44788f61 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS           | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete                |
| NodesDecommissioningTest      | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": false}                    | integration | https://buildkite.com/redpanda/redpanda/builds/77379#019aecf0-a42f-4f66-97db-0524ba4827e7 | FLAKY         | 19/21    | upstream reliability is '93.16239316239316'. current run reliability is '90.47619047619048'. drift is 2.6862 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77607">build#77607</a></summary>


| test_class              | test_method             | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                        |
|:------------------------|:------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest | TestBasicRemoveTopics   |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77607#019b04f7-e94d-4b1e-9f15-4fa9daac43bd | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                     |
| ReplicatedMetastoreTest | TestBasicRemoveTopics   |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77607#019b056f-20b7-4045-a77f-c2662d2e9e1c | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                     |
| ScalingUpTest           | test_fast_node_addition | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77607#019b053b-7691-400f-a3da-8fd35afccb96 | FLAKY         | 29/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0292, p0=0.5889, reject_threshold=0.0100. adj_baseline=0.0851, p1=0.2632, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77631">build#77631</a></summary>


| test_class                       | test_method                           | test_arguments                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:---------------------------------|:--------------------------------------|:-----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest         | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77631#019b067d-b219-4fcc-afc4-1e79670ba2b4 | FLAKY         | 12/21    | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1046, p0=0.0006, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss            | null                               | integration | https://buildkite.com/redpanda/redpanda/builds/77631#019b067d-b216-4018-a739-ea109bb608db | FLAKY         | 18/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0549, p0=0.3012, reject_threshold=0.0100. adj_baseline=0.1558, p1=0.3771, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss    |
</details>
<!--/body-->

dotnwat committed:
ct: calculate L0 max GC eligible epoch

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: rename epoch source to cluster support

This abstraction is no longer restricted to just the epoch source, and
calling it cluster support better reflects is expanded scope.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: provide default implementation of max gc calc

The final top-level calculation of the max GC eligible epoch operates on
the two pure-virtual methods get_partitions[_max_gc_epochs]. This commit
provides that default implementation as part of the cluster_support
interface to make it easy to test: we can still override the top level
calculation, but we can also test the default implementation by
providing different implementations of the function dependencies.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: add tests for l0 gc default cluster support impl

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
cluster: make maximum epoch age configurable

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
cluster: enforce soft limit on cached epoch age

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: add additional logging to L0 GC worker

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: use top-level last applied offset as snapshot epoch

The partition snapshot used to drive L0 GC uses an epoch to identify a
consistent snapshot of state. The snapshot is taken from the topic table
which contains a last applied offset which could be used as an epoch,
but if no changes are happening to the topic table then the epoch
doesn't move. Instead we use the last applied offset of the top-level
mux state machine at the controller level which will move consistently
along with the topic table, but at a non-zero forward pace (in the worst
case the epoch service moves this forward forcefully).

The result without this change is that if no new cloud topics are
created then GC stalls at a non-moving minimum epoch.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: add L0 GC probe

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: add test to verify that GC deletes are occuring

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

oleiman commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+            }
+
+            // Detect concurrent changes to the topic table.
+            topic_table.check_topics_map_stable(iter_start_rev);
```
thought: aside from right here, we only seem to use this in partition_balancer_planner. a comment describing why it's particularly important here might be helpful.

rockwotj replied:
This throws if not stable right? Should we convert to `std::unexpected`?

dotnwat replied:
@oleiman it doesn't have anything to do with semantics of topic table, it's just to avoid accessing invalid iterators since we mix iteration with scheduling points. does this update to the comment help?

```diff
diff --git a/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc b/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
index a6bf5e32f7..39de530fee 100644
--- a/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
+++ b/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
@@ -194,7 +194,8 @@ public:
                 co_return std::unexpected("Abort requested");
             }

-            // Detect concurrent changes to the topic table.
+            // Detect concurrent changes to the topic table to avoid accessing
+            // an invalid iterator.
             topic_table.check_topics_map_stable(iter_start_rev);
         }
```

dotnwat replied:
@rockwotj yes it does throw. fixed to use std::expected.

oleiman replied:
>since we mix iteration with scheduling points

ah yeah missed the `yield`

oleiman commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+         * Below we merge the two result sets and walk the final result back to
+         * account for the partition with the smallest eligible gc epoch.
+         */
+        auto result = cluster_epoch(partitions.value().last_applied());
```
I don't quite understand the relationship between last applied topic table revision and the partition epochs.

dotnwat replied:
Did you read through the "L0 data management" tab in the cloud topics RFC (and especially the appendix describing edge cases) and still have the same question?

I do need to figure out how to document the algorithm in the code more effectively.

oleiman replied:
I think I understand the algorithm. What i missed is that the epoch is fed by the controller committed offset. Do I have that right?

dotnwat replied:
> Do I have that right?

Yep! We should get together and make a document or picture. I think the entire algorithm is pretty simple, but it's a lot of details to track and put together. For that reason I think it's kind of hard to understand in "code form".

oleiman reviewed (commented):
looks pretty sensible to me. just a couple of questions for my own edification.

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.h:
```diff
@@ -180,6 +181,23 @@ class level_zero_gc {
+              model::topic_namespace_eq>;
+
+            partition_map partitions;
+            model::revision_id last_applied;
```
nit: since we're in the epoch_source class here, does it make conceptual sense to cast this to an epoch? It seems intuitive to think about this as the cluster epoch at which the partition snapshot was taken -- is that accurate?

dotnwat replied:
yep. i casted the incoming revisions to cluster epoch as they enter into that sacred area

rockwotj commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+                     * replica. a reported epoch is valid when it is collected,
+                     * independent of the replica's current leadership status.
+                     * we reduce across replicas using `max` because that is the
+                     * most optimistic value we can take.
```
I had a good think about this and it makes sense to use the optimistic value here, but it was immediate obvious to me (although maybe I just need more sleep 😴).

dotnwat replied:
i update the comment to be more verbose in the explanation. i'm not sure it is any clearer, but it is longer and it is differenter

```diff
diff --git a/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc b/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
index a1abc60d6a..90776f6b8b 100644
--- a/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
+++ b/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
@@ -238,11 +238,21 @@ public:
                 const auto& tp_ns = topic_status.first;
                 for (const auto& partition_status : topic_status.second) {
                     /*
-                     * partition_status is the partition health reported by a
-                     * replica. a reported epoch is valid when it is collected,
-                     * independent of the replica's current leadership status.
-                     * we reduce across replicas using `max` because that is the
-                     * most optimistic value we can take.
+                     * calculate the max gc epoch for each partition. the catch
+                     * here is that this value is reported through the health
+                     * reporting system, and that system reports information for
+                     * all partition replicas (leader and followers). so how do
+                     * we know which value to use here? first, the max gc epoch
+                     * only increases in value. second, we recognize that
+                     * all reported values from any replica are valid at (and
+                     * forever after) the moment they are reported. third, only
+                     * the leader advances the epoch.
+                     *
+                     * because of the second point, using max gc epoch from any
+                     * replica will result in correct behavior, however it may
+                     * be pessimistic. using the one from the leader is better,
+                     * but leadership is a lagging signal. instead, we can take
+                     * the maximum reported as the most optimistic value.
                      *
                      * if a replica reports no epoch then it is considered to be
                      * in an indeterminite state and it has no affect on the
```

rockwotj commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+                     * if a replica reports no epoch then it is considered to be
+                     * in an indeterminite state and it has no affect on the
+                     * computed result. this has the side effect / benefit of
+                     * skipping any non-cloud topic partitions.
```
for cloud topic partitions, this means we fallback to the topic revision right?

dotnwat replied:
ahh good question. yeh, i think it's the topic creation revision, but that is handled in the L0 STM not here. I updated the comment to explain the different cases:

```diff
diff --git a/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc b/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
index 90776f6b8b..673a981094 100644
--- a/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
+++ b/src/v/cloud_topics/level_zero/gc/level_zero_gc.cc
@@ -256,8 +256,25 @@ public:
                      *
                      * if a replica reports no epoch then it is considered to be
                      * in an indeterminite state and it has no affect on the
-                     * computed result. this has the side effect / benefit of
-                     * skipping any non-cloud topic partitions.
+                     * computed result (effectively it is treated as having
+                     * epoch 0 in the max reduction across replicas).
+                     *
+                     * if all replicas report no epoch then the partition is not
+                     * included in the result set returned to the caller. this
+                     * covers two cases.
+                     *
+                     * the first case is that the partition is part of a
+                     * standard topic. in this case the partition will also not
+                     * be in the set returned by `get_partitions` and thus the
+                     * join in `max_gc_eligible_epoch` will ignore the topic.
+                     *
+                     * in the second case the join would fail, and later succeed
+                     * in the once at least one replica is returning max gc
+                     * epoch. this shouldn't be a problem in practice: there is
+                     * a narrow window at start-up time where a partition is
+                     * bootstrapping the L0 CT STM state where the state is
+                     * unknown. for brand new partitions this should be the
+                     * partition's creation revision ID.
                      */
```

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+            }
+
+            // Detect concurrent changes to the topic table.
+            topic_table.check_topics_map_stable(iter_start_rev);
```
I generally try to avoid mixing exceptions and result types. This will throw if there's a concurrent update -- can you catch it and return an unexpected result instead?

dotnwat replied:
fixed

rockwotj commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+        const auto iter_start_rev = topic_table.topics_map_revision();
+
+        partitions_snapshot snap;
+        snap.last_applied = topic_table.last_applied_revision();
```
I thought we'd be using the topic's revision in addition to this? Like could a partition be missing from the health report or is that not possible? I'm not sure of the consistency model for health reports TBH

rockwotj replied:
Actually due to caching of a cluster epoch, it's possible a new partitions writes old epochs before it shows up in health reports right.. Is that a problem? Do we need to seed the state machine's min epoch with the revision ID?

dotnwat replied:
> I thought we'd be using the topic's revision in addition to this?

the `topic_table.last_applied_revision();` is referring to a version for the consistent snapshot of all existing cloud topic partitions at this current point in time. that establishes an upper bound on the epoch we'll use for GC.

but it isn't necessarily safe to use that epoch for gc. each of the partitions in the snapshot must then be queried (from health report data) and have an opportunity to lower the final epoch. the consistent of health reports shouldn't matter: what matters is that every partition is queried. if we don't have data from a partition then we must retry the entire calculation because it is unknown how low that partition would want the final epoch to be.

dotnwat replied:
> Actually due to caching of a cluster epoch, it's possible a new partitions writes old epochs before it shows up in health reports right.. Is that a problem?

I think you may be referring to the exact issue that is discussed in the data management section of the RFC called _The epoch-gc-partition-creation race_?

> Do we need to seed the state machine's min epoch with the revision ID

yes, i believe evgeny is reporting something like that from the L0 STM.

rockwotj replied:
> I think you may be referring to the exact issue that is discussed in the data management section of the RFC called The epoch-gc-partition-creation race?

Yes. It's the part about P1 being created with a minimum epoch that I don't know where that happens.

> yes, i believe evgeny is reporting something like that from the L0 STM.

Do you have a code pointer?

dotnwat replied:
> Do you have a code pointer?

@Lazin calculates it in [ctp_stm::estimate_inactive_epoch](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_topics/level_zero/stm/ctp_stm.cc#L169) (cc @wdberkeley)

rockwotj replied:
Hrm so that calls: https://github.com/redpanda-data/redpanda/blob/29b610d90fade0cd430dfa5ec673810ee13a3f24/src/v/cloud_topics/level_zero/stm/ctp_stm_state.cc#L34C12-L34C34

And `_min_epoch_lower_bound` is only set in `advance_epoch` and `advance_last_reconciled_offset`. I guess my mental model is that `_max_applied_epoch` should be set to the partition's revision if the state is for a new partition. never being seeded with the partition's creation epoch however, which is where the disconnect is for me. Maybe we need to chat live about this tomorrow

dotnwat replied:
> partition's revision if the state is for a new partition

yes it should, or something more pessimistic. i recall Evgeny convincing me at some point we were good to go, but now the details are escaping me. @Lazin is this initialization state something we overlooked or it's safe (and if so, how)?

Yeh let's plan to chat about this tomorrow. I pinged you and Evgeny on slack. Pretty sure this is in the design rfc. I recall discussions in ctm_stm reviews regarding handling a partition that transitions to an empty state, but now I'm realizing we may not have discussed handling brand new partitions, and I think we need to apply a minimum (or something that achieves at least as safe a guarantee)

Lazin replied:
The ctp_stm doesn't take topic creation epoch (or initial revision id) into account. The STM is not the best place to do this.

rockwotj commented on src/v/config/configuration.cc:
```diff
@@ -4511,6 +4511,12 @@ configuration::configuration()
+  , epoch_service_max_epoch_age(
+      *this,
+      "epoch_service_max_epoch_age",
+      "The maximum age of an epoch that the epoch service will publish",
```
I still feel like I understand this better as "The interval at which the cluster epoch is incremented"?

dotnwat replied:
yeh, yours is more direct and literal. fixed

rockwotj commented on src/v/config/configuration.cc:
```diff
@@ -4517,6 +4517,13 @@ configuration::configuration()
+      *this,
+      "epoch_service_cached_epoch_age_soft_limit",
+      "When the age of a cached epoch reaches the soft limit it triggers a "
+      "background refresh",
```
I understand this better as "the local cache duration of a cluster wide epoch".

dotnwat replied:
yeh, yours is more direct and literal. fixed

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.h:
```diff
@@ -179,7 +179,7 @@ class level_zero_gc {
      * Interface for computing the maximum epoch eligible for GC.
      */
-    class epoch_source {
+    class cluster_support {
```
I liked epoch_source because it was at least clear that it derived its outputs from epochs (and if we change last_applied to be cluster_epoch, all of its functions would be returning epochs, which makes "source" feel like an apt name). IMO cluster_support doesn't tell users much about what they should expect from this class.

dotnwat replied:
yeh no strong preference. ill change it back.

dotnwat replied:
i think the original thought was that it was supposed to be an abstraction over the epoch service and then its scope increased. but if we dont' think about it is as an abstraction over the epoch service explicitly, then it is still accurate to call it an epoch source!

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -4511,6 +4511,12 @@ configuration::configuration()
       "term storage.",
       {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
       5min)
+  , epoch_service_max_epoch_age(
```
Let's prefix this with "cloud_topics_", and add some color that it applies to cloud topics short term storage epoch

dotnwat replied:
sgtm i think keeping them prefixed is nice. unless @rockwotj has immediate plans for epoch service beyond CT we can rename it and alias it we ever want to use it for something else?

dotnwat replied:
fixed

rockwotj replied:
nope, good to scope to CT only

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+                }
+
+                // this partition may hold back the max GC eligible epoch
+                result = std::min(result, pit->second);
```
I'm kind of surprised we don't have to do any validation of the revision_ids of the partitions. I can't immediately think of a case where this is problematic, but it does seem like non-obvious behavior that the values returned may not have anything to do with the snapshot we took

dotnwat replied:
> but it does seem like non-obvious behavior that the values returned may not have anything to do with the snapshot we took

what makes you say this? epochs and revisions are all sampled as committed offsets of the controller log.

andrwng replied:
The closest "weird" behavior I can think of is a scenario where the health report has some later topic revision than the one in the snapshot (e.g. because we are a stale controller leader). In this scenario, if the health reports were actually referring to a later revision of the topic, and so we may apply an epoch that doesn't make any sense for the partition from the controller. It's technically safe to do so because this older partition _should_ be GCed, but that mismatch seemed worth calling out.

I don't think there is a correctness concern, but it's non-obvious to me because we're operating on pieces of state that are snapshotted non-atomically with respect to one another, and we don't address in comments or otherwise why that's safe.

dotnwat replied:
> where the health report has some later topic revision than the one in the snapshot

to be clear you are referring to weird behavior where two topics with the same name are being compared but they are really different topic instances?

dotnwat replied:
> operating on pieces of state that are snapshotted non-atomically with respect to one another

that's correct. it's why we use lower and upper bounds to control for that non-determisim (appendix in data management section of the rfc).

it's clear from this review that i need to put together a much tighter set of documentation around GC. i'm fairly confident that the GC algorithm is sound. but after talking with bharath (and thinking about your questions here) we do have some very specific requirements around revisions that aren't precisely defined, and i wouldn't be surprised if we need to made some minor adjustments to the derivation of those revisions before GA.

Lazin replied:
I think we need to check if the snapshot that we created by looking at topic table is actually consistent. The controller offset of the snapshot is the upper bound for the GC epoch. Do we enforce this invariant somewhere?

Lazin replied:
Basically, the `get_partitions` method returns a set of partitions and then we're looking at every partition in isolation in this method. But the list returned by `get_partitions` is generated from state created by applying controller log commands up to offset N. This offset N is an upper bound for GC.

oleiman commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+        auto health_report
+          = co_await health_monitor_->local().get_cluster_health(
+            cluster::cluster_report_filter{},
+            cluster::force_refresh::no,
```
> force_refresh::no

because the worst case is a too conservative result, right?

dotnwat replied:
the argument here is that refreshing is kind of expensive (is it? i thought so), but refreshing doesn't buy much: if there is a natural refresh every say 10 minutes, then in principle we are only ever 10 minutes behind in GC. for GC we really care about throughput not that we are deleting data as soon as it becomes eligible.

oleiman replied:
ya that's what I figured. thx

dotnwat committed:
ct: clarify use of concurrent iteration detection

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: convert throw into std::expected

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: clarify comment

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: clarify how missing ct epoch data is handled

Adds cloud_topics prefix to group related configs together, and applies
feedback from review to make the meaning of the configuration clearer.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: improve naming of cloud topics configuration option

Adds to cloud_topics name group, and applies feedback from review to
making the name and description more literal.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: cast incoming revision to cluster epoch

This keeps all of the types inside the GC calculation as epochs even
though the source of the epochs come from controller as revision ids.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ct: revert naming of epoch_source to cluster_support

The people have spoken!

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

micheleRP commented on src/v/config/configuration.cc:
```diff
@@ -4511,6 +4511,18 @@ configuration::configuration()
+  , cloud_topics_epoch_service_epoch_increment_interval(
+      *this,
+      "cloud_topics_epoch_service_epoch_increment_interval",
+      "The interval at which the cluster epoch is incremented",
```
```suggestion
      "The interval at which the cluster epoch is incremented.",
```

micheleRP commented on src/v/config/configuration.cc:
```diff
@@ -4511,6 +4511,18 @@ configuration::configuration()
+  , cloud_topics_epoch_service_local_epoch_cache_duration(
+      *this,
+      "cloud_topics_epoch_service_local_epoch_cache_duration",
+      "The local cache duration of a cluster wide epoch",
```
```suggestion
      "The local cache duration of a cluster wide epoch.",
```

micheleRP commented on src/v/config/configuration.cc:
```suggestion
      "Once experimental features are enabled they cannot be disabled.",
```

dotnwat committed:
config: apply feedback from docs team

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77631-->
<!--b787a8a6-defd-476f-a0c4-ad548b5830cf-->
<!--title-->
## Retry command for [Build#77631](https://buildkite.com/redpanda/redpanda/builds/77631)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_rebalancing_node@{"shutdown_decommissioned":false}
```
<!--/body-->

Lazin commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+            co_await seastar::maybe_yield();
+
+            if (as && as->abort_requested()) {
+                co_return std::unexpected("Abort requested");
```
why are we returning string as an error value here?

dotnwat replied:
Since the error conditions are all self-contained in this GC worker file, defining an enum error code for every condition would be a lot of overhead. I doubt we'll continue to use strings as it grows larger. Does that answer your question or were you asking something else?

oleiman commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -154,7 +154,8 @@ ss::future<> cluster_epoch_service<Clock>::raft0_state::service_loop() {
         try {
             co_await ss::sleep_abortable<Clock>(
-              config::shard_local_cfg().epoch_service_max_epoch_age(),
+              config::shard_local_cfg()
+                .cloud_topics_epoch_service_epoch_increment_interval(),
```
nit: should be in the next commit

vbotbuildovich commented:
/backport v25.3.x

dotnwat commented:
@Lazin 

> I think we need to check if the snapshot that we created by looking at topic table is actually consistent. The controller offset of the snapshot is the upper bound for the GC epoch. Do we enforce this invariant somewhere?

The belief is that it is consistent because updates to the topics table are themselves commands in the controller log. But I agree, we should have more constraints. One simple thing to do is sample before and after taking the snapshot and makes ure it didn't change. Kinda like a sequence lock?
